src/Mvc/test/Mvc.IntegrationTests/ValidationWithRecordIntegrationTests.cs

```diff
@@ -2298,6 +2298,119 @@ public async Task Validation_InifnitelyRecursiveModel_ValidationOnTopLevelParame
 
         private static void Validation_InifnitelyRecursiveModel_ValidationOnTopLevelParameterMethod([Required] RecursiveModel model) { }
 
+        [Fact]
+        public async Task Validation_ValidatorsDefinedOnRecordTypeProperties()
+        {
+            // Arrange
+            var modelType = typeof(RecordTypeWithValidatorsOnProperties);
+            var modelMetadataProvider = TestModelMetadataProvider.CreateDefaultProvider();
+            var modelMetadata = modelMetadataProvider.GetMetadataForType(modelType);
+            var parameterBinder = ModelBindingTestHelper.GetParameterBinder(modelMetadataProvider);
+            var expected = $"Record type '{modelType}' has validation metadata defined on property 'Property1' that will be ignored. " +
+                "Consider specifying these on the constructor parameter 'Property1' instead.";
+
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "parameter",
+                ParameterType = modelType,
+            };
+
+            var testContext = ModelBindingTestHelper.GetTestContext(request =>
+            {
+                request.QueryString = new QueryString("?Property1=8");
+            });
+
+            var modelState = testContext.ModelState;
+
+            // Act & Assert
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(() =>
+                parameterBinder.BindModelAsync(parameter, testContext, modelMetadataProvider, modelMetadata));
+
+            Assert.Equal(expected, ex.Message);
+        }
+
+        private record RecordTypeWithValidatorsOnProperties(string Property1)
```

Comment:
Nit: Move this above `Validation_ValidatorsDefinedOnRecordTypeProperties()`. Same for the other two new `record` types and the methods that use them. (We either place inner types at the bottom or just above where they're used.)